lib/rules/yoda.js

```diff
@@ -248,7 +248,7 @@ module.exports = {
                 return (node.operator === "&&" &&
                     (leftLiteral = getNormalizedLiteral(left.left)) &&
                     (rightLiteral = getNormalizedLiteral(right.right, Number.POSITIVE_INFINITY)) &&
-                    leftLiteral.value <= rightLiteral.value &&
+                    (typeof leftLiteral.value === "string" || leftLiteral.value <= rightLiteral.value) &&
```

Comment:
This would introduce false negatives:

```js
/*eslint yoda: ["error", "never", { "exceptRange": true }]*/

if ("b" <= x && x < "a") {
    // ...
}
```

Looks like we should rather replace the `Number.POSITIVE_INFINITY` logic with something else?

Reply:
> This would introduce false negatives:
> 
> ```js
> /*eslint yoda: ["error", "never", { "exceptRange": true }]*/
> 
> if ("b" <= x && x < "a") {
>     // ...
> }
> ```
> 


Should I give a check for rightLiteral as well ? like this 

```js
(typeof rightLiteral.value === "string" || rightLiteral.value <= leftLiteral.value)
```

??

Reply:
> Looks like we should rather replace the `Number.POSITIVE_INFINITY` logic with something else?

I would give a range options with `Number.MAX_SAFE_INTEGER` and `Number.MIN_SAFE_INTEGER`

Reply:
The problem is: `"b" <= x && x < "a"` should be reported as an error, similar to how the rule already reports `2 <= x && x < 1`. Both are invalid range checks, because the right value isn't greater than the left value.

The issue is when one of these values is unknown (e.g., an identifier `foo`).

The actual logic handles this by treating unknown values as positive or negative infinity values, so the comparison works well for numbers, but it fails for strings.

Reply:
Example from #12883:

```js
/* eslint yoda: ["error", "never", { exceptRange: true }] */

if (1 < x && x < MAX ) {} // no error

if ("a" < x && x < MAX ) {} // error on "a" < x
```

The different behavior is caused by:

```js
1 < Number.POSITIVE_INFINITY; // true

"a" < Number.POSITIVE_INFINITY; // false
```

Reply:
I am pretty sure js engine uses the product of char's ASCII value their position for comparing strings like [`they are compared letter-by-letter`](https://javascript.info/comparison#string-comparison).

![image](https://user-images.githubusercontent.com/26347874/76969796-e6e09700-6950-11ea-8cf6-e71478184044.png)

so for this 

` ("b" <= x && x < "a")`, we can compare the left right literals with this algorithm
 

but not sure about this as they are of different data types. So not sure what to do for this. 
It is either to allow them or not !

`("a" < x && x < MAX )`

Reply:
`("a" < x && x < MAX )` should be allowed (it isn't in the actual master version), but not in a way that would also allow `("b" <= x && x < "a")` which correctly isn't allowed in the actual master version.

Reply:
> but not in a way that would also allow `("b" <= x && x < "a")` which correctly isn't allowed in the actual master version.


@mdjermanovic  can we use `eval` for `("b" <= x && x < "a")` (when both left and right literals are `string`) ?

![image](https://user-images.githubusercontent.com/26347874/77049455-1d202400-69ee-11ea-948b-127e60482eb1.png)


also just to summarise few things  (let me know which point is wrong)

when
- **`rightLiteral` is `string` but not `leftLiteral`:** 

  it should allow (it does in master [`demo`](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IHlvZGE6IFtcImVycm9yXCIsIFwibmV2ZXJcIiwgeyBleGNlcHRSYW5nZTogdHJ1ZSB9XSAqL1xuXG5cbmlmICggeCA8IE1BWCAmJiB4IDwgXCJhXCIgKSB7fSAiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=))
  no changes require for this

- **`leftLiteral` is  `string` but not `rightLiteral` is not a `string`:** 
  it should allow but in `master` its not allowing [`demo`](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IHlvZGE6IFtcImVycm9yXCIsIFwibmV2ZXJcIiwgeyBleGNlcHRSYW5nZTogdHJ1ZSB9XSAqL1xuXG5cbmlmIChcImFcIiA8IHggJiYgeCA8IE1BWCApIHt9IC8vIGVycm9yIG9uIFwiYVwiIDwgeCIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==) and this is from the original issue
  allowing in this PR

- **when both `leftLiteral` and `rightLiteral` are `string`:** 
   it should compare them and then allow or disallow based on the result  solution I proposed is to use 
   ```js
    eval(`${leftLiteral} ${operator} ${rightLiteral}`)
    ```

Reply:
> * **`rightLiteral` is `string` but not `leftLiteral`:**
>   it should allow (it does in master [`demo`](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IHlvZGE6IFtcImVycm9yXCIsIFwibmV2ZXJcIiwgeyBleGNlcHRSYW5nZTogdHJ1ZSB9XSAqL1xuXG5cbmlmICggeCA8IE1BWCAmJiB4IDwgXCJhXCIgKSB7fSAiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=))
>   no changes require for this

Example in the demo is lint-free because there are no "yoda" conditions. It would be lint-free with `{ exceptRange: false }`, too.

An example where `rightLiteral` is string and `leftLiteral` isn't string would be something like:

```js
if (MIN < x && x < "a" ) {}
```

This also isn't error with `"never"`, regardless of `exceptRange`. It would be error with `"always"`, and that's point `1.  Possible inconsistency between "never" and "always"` from #12883 (it isn't a problem with strings only, numbers would also cause error).

> * **`leftLiteral` is  `string` but not `rightLiteral` is not a `string`:**
>   it should allow but in `master` its not allowing [`demo`](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IHlvZGE6IFtcImVycm9yXCIsIFwibmV2ZXJcIiwgeyBleGNlcHRSYW5nZTogdHJ1ZSB9XSAqL1xuXG5cbmlmIChcImFcIiA8IHggJiYgeCA8IE1BWCApIHt9IC8vIGVycm9yIG9uIFwiYVwiIDwgeCIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==) and this is from the original issue
>   allowing in this PR

:+1: that's point `2.` from #12883.

> * **when both `leftLiteral` and `rightLiteral` are `string`:**
>   it should compare them and then allow or disallow based on the result  solution I proposed is to use
>   ```js
>    eval(`${leftLiteral} <= ${rightLiteral}`)
>   ```

Why is `eval` necessary here, `leftLiteral.value <= rightLiteral.value` already works well if both literals are strings or any other literals.

Reply:
cc @mdjermanovic can you check the latest commit I did some changes to fix both the issues mentioned in the original issue.